File: app/src/main/java/com/commonpepper/photosen/Photosen.java
package com.commonpepper.photosen;

import android.app.Activity;
import android.app.Application;
import android.content.pm.PackageManager;
import android.content.res.TypedArray;
import android.os.Build;

import com.commonpepper.photosen.database.PhotosenDatabase;
import com.commonpepper.photosen.network.KeyFormatInterceptor;
import com.commonpepper.photosen.network.FlickrApi;
import com.crashlytics.android.Crashlytics;
import com.crashlytics.android.core.CrashlyticsCore;
import com.google.firebase.analytics.FirebaseAnalytics;

import androidx.core.app.ActivityCompat;
import androidx.room.Room;

import io.fabric.sdk.android.Fabric;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class Photosen extends Application {

    public static final String API_URL = "https://api.flickr.com/";

    public static final int PAGE_SIZE = 30;
    public static final int PREFETCH_DISTANCE = 5;
    public static final String PACKAGE_NAME = "com.commonpepper.photosen";
    public static final String PREFERENCES = PACKAGE_NAME + ".PREFERENCES";

    private static FirebaseAnalytics firebaseAnalytics;
    private static Photosen instance;
    private PhotosenDatabase database;

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;
        firebaseAnalytics = FirebaseAnalytics.getInstance(this);
        configureCrashReporting();

        database = Room.databaseBuilder(this, PhotosenDatabase.class, "photosen_database")
                .build();
    }

    private void configureCrashReporting() {
        CrashlyticsCore crashlyticsCore = new CrashlyticsCore.Builder()
                .disabled(BuildConfig.DEBUG)
                .build();
        Fabric.with(this, new Crashlytics.Builder().core(crashlyticsCore).build());
    }

    public static Photosen getInstance() {
        return instance;
    }

    public PhotosenDatabase getDatabase() {
        return database;
    }

    public static FirebaseAnalytics getFirebaseAnalytics() {
        return firebaseAnalytics;
    }

    public static FlickrApi getFlickrApi() {
        return new Retrofit.Builder()
                .baseUrl(API_URL)
                .client(getClient())
                .addConverterFactory(GsonConverterFactory.create())
                .build()
                .create(FlickrApi.class);
    }

    private static OkHttpClient getClient() {
        return new OkHttpClient.Builder()
                .addInterceptor(new KeyFormatInterceptor())
                .build();
    }

    public static boolean isStoragePermissionGranted(Activity activity) {
        if (Build.VERSION.SDK_INT >= 23) {
            if (ActivityCompat.checkSelfPermission(activity, android.Manifest.permission.WRITE_EXTERNAL_STORAGE)
                    == PackageManager.PERMISSION_GRANTED) {
                return true;
            } else {
                ActivityCompat.requestPermissions(activity, new String[]{android.Manifest.permission.WRITE_EXTERNAL_STORAGE}, 1);
                return false;
            }
        } else {
            return true;
        }
    }
}

File: app/src/main/java/com/commonpepper/photosen/Photosen.kt
package com.commonpepper.photosen

import android.Manifest.permission
import android.app.Activity
import android.app.Application
import android.content.pm.PackageManager
import android.os.Build.VERSION
import androidx.core.app.ActivityCompat
import androidx.room.Room
import com.commonpepper.photosen.database.PhotosenDatabase
import com.commonpepper.photosen.network.BooleanTypeAdapter
import com.commonpepper.photosen.network.FlickrApi
import com.commonpepper.photosen.network.KeyFormatInterceptor
import com.google.firebase.FirebaseApp
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.gson.GsonBuilder
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory


class Photosen : Application() {
    lateinit var database: PhotosenDatabase

    override fun onCreate() {
        super.onCreate()
        instance = this
        FirebaseApp.initializeApp(this)
        firebaseAnalytics = FirebaseAnalytics.getInstance(this)
        FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(true)
        firebaseAnalytics.setAnalyticsCollectionEnabled(true)
        database = Room.databaseBuilder(this, PhotosenDatabase::class.java, "photosen_database")
                .fallbackToDestructiveMigration()
                .build()
    }

    companion object {
        private const val API_URL = "https://api.flickr.com/"
        const val PAGE_SIZE = 30
        const val PREFETCH_DISTANCE = 5
        const val PACKAGE_NAME = "com.commonpepper.photosen"
        const val PREFERENCES = "$PACKAGE_NAME.PREFERENCES"
        lateinit var firebaseAnalytics: FirebaseAnalytics
        lateinit var instance: Photosen

        val flickrApi: FlickrApi by lazy {
            val builder = GsonBuilder()
            builder.registerTypeAdapter(Boolean::class.java, BooleanTypeAdapter())
            val gson = builder.create()
            Retrofit.Builder()
                    .baseUrl(API_URL)
                    .client(client)
                    .addConverterFactory(GsonConverterFactory.create(gson))
                    .build()
                    .create(FlickrApi::class.java)
        }

        private val client by lazy {
            OkHttpClient.Builder()
                    .addInterceptor(KeyFormatInterceptor())
                    .build()
        }

        fun isStoragePermissionGranted(activity: Activity?): Boolean {
            return if (VERSION.SDK_INT >= 23) {
                if (ActivityCompat.checkSelfPermission(activity!!, permission.WRITE_EXTERNAL_STORAGE)
                        == PackageManager.PERMISSION_GRANTED) {
                    true
                } else {
                    ActivityCompat.requestPermissions(activity, arrayOf<String?>(permission.WRITE_EXTERNAL_STORAGE), 1)
                    false
                }
            } else {
                true
            }
        }
    }
}